microservice/package/src/main/java/com/cumulocity/agent/packaging/microservice/impl/MicroserviceDockerClientImpl.java

```diff
@@ -1,81 +1,261 @@
+import lombok.extern.slf4j.Slf4j;
 import org.apache.commons.io.IOUtils;
+import org.apache.maven.MavenExecutionException;
+import org.apache.maven.plugins.annotations.Parameter;
 import org.codehaus.plexus.component.annotations.Component;
 import org.codehaus.plexus.logging.AbstractLogEnabled;
 import org.codehaus.plexus.personality.plexus.lifecycle.phase.Startable;
 import org.codehaus.plexus.personality.plexus.lifecycle.phase.StartingException;
 import org.codehaus.plexus.personality.plexus.lifecycle.phase.StoppingException;
 
-import java.io.File;
-import java.io.FileOutputStream;
-import java.io.IOException;
-import java.io.InputStream;
-import java.util.List;
+import java.io.*;
+import java.time.Duration;
+import java.util.*;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicBoolean;
 
-import static com.spotify.docker.client.DockerClient.ListImagesParam.byName;
 
 @Component(role = MicroserviceDockerClient.class)
+@Slf4j(topic = "MicroserviceDockerClient")
 public class MicroserviceDockerClientImpl extends AbstractLogEnabled implements MicroserviceDockerClient, Startable {
-    private DockerClient docker;
 
-    public void saveDockerImage(final String image, final File targetFile) throws InterruptedException, DockerException, IOException {
-        try (
-                final InputStream inputStream = docker.save(image);
-                final FileOutputStream outputStream = new FileOutputStream(targetFile)
-        ) {
+    DockerClient dockerClient;
+
+    @SneakyThrows
+    public void buildDockerImage(String dockerDirectory, Set<String> tags, Map<String, String> buildArgs, String networkMode, Integer dockerBuildTimeout) {
+
+        BuildImageCmd buildImageCmd = dockerClient.buildImageCmd(new File(dockerDirectory)).withTags(tags);
+
+        for (Map.Entry<String, String> buildArgument : buildArgs.entrySet()) {
+            buildImageCmd = buildImageCmd.withBuildArg(buildArgument.getKey(), buildArgument.getValue());
+        }
+
+        if (!Strings.isNullOrEmpty(networkMode)) {
+            buildImageCmd = buildImageCmd.withNetworkMode(networkMode);
+        }
+
+        log.info("Building Docker image. Docker dir={},tags={}, build arguments={}", dockerDirectory, tags, buildArgs);
+        ImageBuildCompletionWaiter imageBuildCompletionWaiter = new ImageBuildCompletionWaiter();
+        buildImageCmd.exec(imageBuildCompletionWaiter);
+        log.info("Waiting for image build to complete (timeout={}s)", dockerBuildTimeout);
+        imageBuildCompletionWaiter.awaitWithTimeout(dockerBuildTimeout);
+
+        log.info("Image build success");
+
+
+    }
+
+    protected class ImageBuildCompletionWaiter extends BuildImageResultCallback {
+
+        String completedImageId;
+        boolean buildFailed;
+        private Future f;
+
+        public void awaitWithTimeout(int seconds) throws MavenExecutionException {
+
+            ExecutorService executorService = Executors.newSingleThreadExecutor();
```

Comment:
You should shutdown the thread pool or use common thread pool as now it may cause memory leak for large build

Reply:
Resolved with last commit